### PR TITLE
Add StandardGraphBuilderPlugins for standalone analysis

### PR DESCRIPTION
**Pull request by ziyilin**
Register invocation plugins for standalone analysis.

This PR should be merged after PR https://github.com/oracle/graal/pull/5894, otherwise the tests will fail. 
This PR is part of https://github.com/oracle/graal/pull/4375.

### CODE REVIEW

File: substratevm/src/com.oracle.graal.pointsto.standalone.test/src/com/oracle/graal/pointsto/standalone/test/ArrayNewInstanceReachabilityTest.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+package com.oracle.graal.pointsto.standalone.test;
+
+import org.junit.Test;
+
+import java.lang.reflect.Array;
+
+/**
+ * This test verifies whether
+ * {@link org.graalvm.compiler.replacements.StandardGraphBuilderPlugins#registerInvocationPlugins}
+ * has been invoked.
```
Comment (cstancu) [inline substratevm/src/com.oracle.graal.pointsto.standalone.test/src/com/oracle/graal/pointsto/standalone/test/ArrayNewInstanceReachabilityTest.java:36]: Please add some details about how this actually asserts that `StandardGraphBuilderPlugins#registerInvocationPlugins` was applied? Which invocation plugin exactly would change the result of this test if not applied?

File: substratevm/src/com.oracle.graal.pointsto.standalone/src/com/oracle/graal/pointsto/standalone/PointsToAnalyzer.java [hunk 2]
```diff
@@ -197,6 +201,14 @@ private PointsToAnalyzer(String mainEntryClass, OptionValues options) {
             NoClassInitializationPlugin classInitializationPlugin = new NoClassInitializationPlugin();
             plugins.setClassInitializationPlugin(classInitializationPlugin);
             aProviders.setGraphBuilderPlugins(plugins);
+            registerInvocationPlugins(originalProviders.getSnippetReflection(), plugins.getInvocationPlugins(), originalProviders.getReplacements(),
+                            false, true, false, originalProviders.getLowerer());
+            /** Disable inlineBeforeAnalysis for two reasons:
+             *  1) It is necessary for standalone analysis;
+             *  2) Standalone analysis uses {@link com.oracle.graal.pointsto.phases.InlineBeforeAnalysisPolicy#NO_INLINING}
+             *  as inline policy, but it's not well supported since https://github.com/oracle/graal/commit/cc2ad502c6877a2c11685491c6609f0ff8ce7b14.
+             **/
```
Comment (cstancu) [inline substratevm/src/com.oracle.graal.pointsto.standalone/src/com/oracle/graal/pointsto/standalone/PointsToAnalyzer.java:210]: How does that commit break `NO_INLINING` policy? cc: @christianwimmer

Comment (ziyilin) [reply substratevm/src/com.oracle.graal.pointsto.standalone/src/com/oracle/graal/pointsto/standalone/PointsToAnalyzer.java:212]: For the changes below:
image
Given a `caller` with the following states: 
image

As the `caller` is not null but `caller.caller` is null. Before change, it goes to the line 77 on the left side. After change, it goes to the line 76 on the right side. When the `policy` here is `InlineBeforeAnalysisPolicy#NO_INLINING`, it finally goes to `InlineBeforeAnalysis#openCalleeScope` which throws an unimplemented error.

**Comment by ziyilin** [comment]
@cstancu Any review updates?

File: substratevm/src/com.oracle.graal.pointsto.standalone.test/src/com/oracle/graal/pointsto/standalone/test/StandardGraphPluginTest.java [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+     * Check the invocation of method {@code StandardGraphBuilderPlugins#registerArrayPlugins(InvocationPlugins, Replacements)}
+     * which is invoked by {@link StandardGraphBuilderPlugins#registerInvocationPlugins}.
+     * @throws NoSuchMethodException
+     */
+    @Test
+    public void testArrayNewInstanceReachability() throws NoSuchMethodException {
+        PointstoAnalyzerTester tester = new PointstoAnalyzerTester(ArrayNewInstancePluginCase.class);
+        tester.setAnalysisArguments(tester.getTestClassName(),
+                        "-H:AnalysisTargetAppCP=" + tester.getTestClassJar());
+        tester.setExpectedReachableMethods(ArrayNewInstancePluginCase.C.class.getDeclaredMethod("foo"));
```
Comment (cstancu) [inline substratevm/src/com.oracle.graal.pointsto.standalone.test/src/com/oracle/graal/pointsto/standalone/test/StandardGraphPluginTest.java:50]: So how this actually check that `org.graalvm.compiler.replacements.StandardGraphBuilderPlugins#registerArrayPlugins` was applied? You could check that `java.lang.reflect.Array#newArray` is not reachable.

Comment (ziyilin) [reply substratevm/src/com.oracle.graal.pointsto.standalone.test/src/com/oracle/graal/pointsto/standalone/test/StandardGraphPluginTest.java:50]: I have added the assertion that `java.lang.reflect.Array#newArray` is not reachable.
The `ArrayNewInstancePluginCase.C#foo` should be reachable whenever `StandardGraphBuilderPlugins#registerArrayPlugins` is applied. But the test can show it is not the case. When I comment out the following code from `StandardGraphBuilderPlugins#registerArrayPlugins`, the reachable assertion of `ArrayNewInstancePluginCase.C#foo` fails.
```
r.register(new InvocationPlugin("newInstance", Class.class, int.class) {
            @Override
            public boolean apply(GraphBuilderContext b, ResolvedJavaMethod targetMethod, Receiver unused, ValueNode componentType, ValueNode length) {
                ValueNode componentTypeNonNull = b.nullCheckedValue(componentType);
                ValueNode lengthPositive = b.maybeEmitExplicitNegativeArraySizeCheck(length);
                b.addPush(JavaKind.Object, new DynamicNewArrayNode(componentTypeNonNull, lengthPositive, true));
                return true;
            }
        });
```

Looks like the `LoadIndexedTypeFlow` is never updated. Maybe it reveals some other bug.

Comment (cstancu) [reply substratevm/src/com.oracle.graal.pointsto.standalone.test/src/com/oracle/graal/pointsto/standalone/test/StandardGraphPluginTest.java:57]: Well of course if you disable the plugin for `Array.newInstance` then there's no `DynamicNewArrayNode` inserted. `java.lang.reflect.Array#newArray` is a native method that is not linked in.

File: substratevm/src/com.oracle.graal.pointsto.standalone/src/com/oracle/graal/pointsto/standalone/PointsToAnalyzer.java (hunk 2 above)
Comment (cstancu) [reply substratevm/src/com.oracle.graal.pointsto.standalone/src/com/oracle/graal/pointsto/standalone/PointsToAnalyzer.java:213]: @ziyilin I'm not sure that I understand what the problem is. Can you provide a patch that fixes `InlineBeforeAnalysisPolicy#NO_INLINING`?

Comment (ziyilin) [reply substratevm/src/com.oracle.graal.pointsto.standalone/src/com/oracle/graal/pointsto/standalone/PointsToAnalyzer.java]: This problem is fixed in the latest master. I have deleted the option setting code.

File: substratevm/src/com.oracle.graal.pointsto.standalone.test/src/com/oracle/graal/pointsto/standalone/test/StandardGraphPluginTest.java [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+     * @throws NoSuchMethodException
+     */
+    @Test
+    public void testArrayNewInstanceReachability() throws NoSuchMethodException {
+        PointstoAnalyzerTester tester = new PointstoAnalyzerTester(ArrayNewInstancePluginCase.class);
+        tester.setAnalysisArguments(tester.getTestClassName(),
+                        "-H:AnalysisTargetAppCP=" + tester.getTestClassJar());
+        // Somehow this method is not reachable if {@code
+        // StandardGraphBuilderPlugins#registerArrayPlugins(InvocationPlugins, Replacements)} is not
+        // invoked.
```
Comment (cstancu) [inline substratevm/src/com.oracle.graal.pointsto.standalone.test/src/com/oracle/graal/pointsto/standalone/test/StandardGraphPluginTest.java:56]: Please remove or update the above comment according to my explanation.

**Review by cstancu — CHANGES_REQUESTED** [review]

File: substratevm/src/com.oracle.graal.pointsto.standalone.test/src/com/oracle/graal/pointsto/standalone/test/StandardGraphPluginTest.java (hunk 4 above)
Comment (ziyilin) [reply substratevm/src/com.oracle.graal.pointsto.standalone.test/src/com/oracle/graal/pointsto/standalone/test/StandardGraphPluginTest.java]: Fixed by removing the comments.

**Review by cstancu — APPROVED** [review]

**Comment by ziyilin** [comment]
@oubidar-Abderrahim Can you merge this PR please? Thank you.

**Comment by zakkak** [comment]
Implements part of https://github.com/oracle/graal/issues/3418

**Pull request merged**